src/table.mjs

```diff
@@ -231,3 +271,91 @@ function likeOperand(operand) {
   return {...operand, value: `%${operand.value}%`};
 }
 
+// This function applies table cell operations to an in-memory table (array of objects).
+export function __table(source, operations) {
+  let {schema, columns} = source;
+  for (const {type, operands} of operations.filter) {
+    const column = operands.find(({type}) => type === "column").value;
+    const resolved = operands.filter(({type}) => type === "resolved");
+    switch (type) {
+      case "eq": {
+        const [{value}] = resolved;
+        source = source.filter((d) => d[column] === value);
+        break;
+      }
+      case "ne": {
+        const [{value}] = resolved;
+        source = source.filter((d) => d[column] !== value);
+        break;
+      }
+      case "c": {
+        const [{value}] = resolved;
+        source = source.filter(
+          (d) => typeof d[column] === "string" && d[column].includes(value)
+        );
+        break;
+      }
+      case "nc": {
+        const [{value}] = resolved;
+        source = source.filter(
+          (d) => typeof d[column] === "string" && !d[column].includes(value)
+        );
+        break;
+      }
+      case "in": {
+        const values = new Set(resolved.map(({value}) => value));
+        source = source.filter((d) => values.has(d[column]));
+        break;
+      }
+      case "nin": {
+        const values = new Set(resolved.map(({value}) => value));
+        source = source.filter((d) => !values.has(d[column]));
+        break;
+      }
+      case "n": {
+        source = source.filter((d) => d[column] == null);
+        break;
+      }
+      case "nn": {
+        source = source.filter((d) => d[column] != null);
+        break;
+      }
+      case "lt": {
+        const [{value}] = resolved;
+        source = source.filter((d) => d[column] < value);
+        break;
+      }
+      case "gt": {
+        const [{value}] = resolved;
+        source = source.filter((d) => d[column] > value);
+        break;
+      }
+      default:
+        throw new Error(`unknown filter type: ${type}`);
+    }
+  }
+  for (const {column, direction} of reverse(operations.sort)) {
+    const compare = direction === "desc" ? descending : ascending;
+    source.sort((a, b) => compare(a[column], b[column]));
+  }
+  if (operations.slice) {
+    source = source.slice(
+      operations.slice.from ?? 0,
+      operations.slice.to ?? Infinity
+    );
+  }
+  if (operations.select?.columns) {
+    if (schema) {
+      const schemaByName = new Map(schema.map((s) => [s.name, s]));
+      schema = operations.select.columns.map((c) => schemaByName.get(c));
+    } else if (columns) {
+      columns = operations.select.columns;
+    }
+    source = source.map((d) =>
+      Object.fromEntries(operations.select.columns.map((c) => [c, d[c]]))
+    );
+  }
+  if (schema) source.schema = schema;
+  else if (columns) source.columns = columns;
```

Comment:
We should move this assignment to within the if (operations.select.columns) block; otherwise, we should allow the source’s existing schema and columns to pass through unaffected, and we shouldn’t mutate the passed-in source (i.e., we only want to assign these properties after creating a copy of source by calling source.map).